lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -132,6 +132,10 @@ static bool isDuplicatableNullaryExpression(Operation *op) {
       return true;
   }
 
+  // Always duplicate XMRs into their use site.
+  if (isa<XMRRefOp>(op))
+    return true;
```

Comment:
Is it required? I think `isExpressionAlwaysInline` is doing the same thing and also this code path is actually dead.

Reply:
I was surprised around this one, too. This is necessary (or it fixes a crash) when running the end-to-end `Wire.fir` integration test with `--disable-opt`. E.g., without this I can produce a crash with:

```
# lldb firtool -- --annotation-file ../test/Dialect/FIRRTL/SFCTests/GrandCentralInterfaces/Wire.fir.anno.json ../test/Dialect/FIRRTL/SFCTests/GrandCentralInterfaces/Wire.fir --disable-opt
(lldb) target create "firtool"
Current executable set to 'firtool' (x86_64).
(lldb) settings set -- target.run-args  "--annotation-file" "../test/Dialect/FIRRTL/SFCTests/GrandCentralInterfaces/Wire.fir.anno.json" "../test/Dialect/FIRRTL/SFCTests/GrandCentralInterfaces/Wire.fir" "--disable-opt"
(lldb) break set --file ExportVerilog.cpp --line 137                                                                                                                                                                                          
Breakpoint 1: where = firtool`isDuplicatableNullaryExpression(mlir::Operation*) + 223 at ExportVerilog.cpp:137:5, address = 0x00000001002860af
(lldb) r
Process 70128 launched: '/Users/schuylere/repos/github.com/llvm/circt/build/bin/firtool' (x86_64)
Process 70128 stopped
* thread #2, stop reason = breakpoint 1.1
    frame #0: 0x00000001002860af firtool`isDuplicatableNullaryExpression(op=0x00006000021060e0) at ExportVerilog.cpp:137:5
   134   
   135     // Always duplicate XMRs into their use site.
   136     if (isa<XMRRefOp>(op))
-> 137       return true;
   138   
   139     // If this is a macro reference without side effects, allow duplication.
   140     if (isa<MacroRefExprOp>(op))
Target 0: (firtool) stopped.
(lldb) bt
* thread #2, stop reason = breakpoint 1.1
  * frame #0: 0x00000001002860af firtool`isDuplicatableNullaryExpression(op=0x00006000021060e0) at ExportVerilog.cpp:137:5
    frame #1: 0x0000000100282387 firtool`isDuplicatableExpression(op=0x00006000021060e0) at ExportVerilog.cpp:151:12
    frame #2: 0x00000001002822a3 firtool`circt::ExportVerilog::isExpressionEmittedInline(op=0x00006000021060e0, options=0x00007ff7bfefdb48) at ExportVerilog.cpp:696:41
    frame #3: 0x000000010029dc8a firtool`(anonymous namespace)::NameCollector::collectNames(this=0x000070000d05cbb0, block=0x0000600002106800) at ExportVerilog.cpp:2866:5
    frame #4: 0x000000010029d6a6 firtool`(anonymous namespace)::StmtEmitter::collectNamesAndCalculateDeclarationWidths(this=0x000070000d05ce08, block=0x0000600002106800) at ExportVerilog.cpp:4532:13
    frame #5: 0x000000010029d5b3 firtool`(anonymous namespace)::StmtEmitter::emitStatementBlock(this=0x000070000d05cd38)::$_30::operator()() const at ExportVerilog.cpp:4555:7
    frame #6: 0x000000010029d535 firtool`decltype(__f=0x000070000d05cd38)::StmtEmitter::emitStatementBlock(mlir::Block&)::$_30>(fp)()) std::__1::__invoke<(anonymous namespace)::StmtEmitter::emitStatementBlock(mlir::Block&)::$_30>((anonymous namespace)::StmtEmitter::emitStatementBlock(mlir::Block&)::$_30&&) at type_traits:3918:1
    frame #7: 0x000000010029d45d firtool`std::__1::invoke_result<(anonymous namespace)::StmtEmitter::emitStatementBlock(mlir::Block&)::$_30>::type std::__1::invoke<(anonymous namespace)::StmtEmitter::emitStatementBlock(__f=0x000070000d05cd38)::$_30>((anonymous namespace)::StmtEmitter::emitStatementBlock(mlir::Block&)::$_30&&) at invoke.h:93:12
    frame #8: 0x000000010029d3e6 firtool`auto circt::pretty::TokenStream<circt::pretty::PrettyPrinter>::scopedBox<circt::pretty::PP, (anonymous namespace)::StmtEmitter::emitStatementBlock(this=0x000070000d05ce10, t=0x000070000d05cd4c, c=0x000070000d05cd38, close=Token @ 0x000070000d05ccf0)::$_30>(circt::pretty::PP&&, (anonymous namespace)::StmtEmitter::emitStatementBlock(mlir::Block&)::$_30&&, circt::pretty::Token) at PrettyPrinterHelpers.h:336:12
    frame #9: 0x000000010028e944 firtool`(anonymous namespace)::StmtEmitter::emitStatementBlock(this=0x000070000d05ce08, body=0x0000600002106800) at ExportVerilog.cpp:4543:6
    frame #10: 0x000000010028d261 firtool`(anonymous namespace)::ModuleEmitter::emitHWModule(this=0x000070000d05d5c0, module=HWModuleOp @ 0x000070000d05d248) at ExportVerilog.cpp:5045:29
    frame #11: 0x000000010028c6c4 firtool`auto emitOperation(this=0x000070000d05d810, op=HWModuleOp @ 0x000070000d05d778)::VerilogEmitterState&, mlir::Operation*)::$_100::operator()<circt::hw::HWModuleOp>(circt::hw::HWModuleOp) const at ExportVerilog.cpp:5278:61
    frame #12: 0x000000010028c31b firtool`llvm::TypeSwitch<mlir::Operation*, void>& llvm::TypeSwitch<mlir::Operation*, void>::Case<circt::hw::HWModuleOp, emitOperation(this=0x000070000d05d818, caseFn=0x000070000d05d810)::VerilogEmitterState&, mlir::Operation*)::$_100>(emitOperation((anonymous namespace)::VerilogEmitterState&, mlir::Operation*)::$_100&&) at TypeSwitch.h:169:7
    frame #13: 0x0000000100284745 firtool`emitOperation(state=0x000070000d05d928, op=0x00006000026085a0)::VerilogEmitterState&, mlir::Operation*) at ExportVerilog.cpp:5278:8
    frame #14: 0x0000000100303e76 firtool`circt::ExportVerilog::SharedEmitterState::emitOps(this=0x00007ff7bfefd5b8, stringOrOp=0x0000000108408f40)::$_45::operator()(circt::ExportVerilog::StringOrOpToEmit&) const at ExportVerilog.cpp:5342:5
```

For this test, the logic for whether or not this is emitted inline is using `isExpressionEmittedInline` whereas with optimizations it is on `isExpressionAlwaysInline`. I didn't dig too deeply, but I expect there is some duplication inherent here that can be cleaned up.

Reply:
Hmm, I see. I agree that we need to refactor that. I'll take a look at later.